core/jvm/src/SegmentPool.kt

```diff
@@ -83,55 +142,148 @@ internal actual object SegmentPool {
     actual fun take(): Segment {
         val firstRef = firstRef()
 
-        val first = firstRef.getAndSet(LOCK)
-        when {
-            first === LOCK -> {
-                // We didn't acquire the lock. Don't take a pooled segment.
-                return Segment.new()
-            }
+        while (true) {
+            when (val first = firstRef.getAndSet(LOCK)) {
+                LOCK -> {
+                    // We didn't acquire the lock. Let's try again
+                    continue
+                }
+
+                null -> {
+                    // We acquired the lock but the pool was empty. Unlock and return a new segment.
```

Comment:
This comment contradicts the flow -- we unlock the bucket *and maybe fall through* into the second-level cache. 
It is guarded by the lock on its own, but still a ground to make mistakes

Reply:
Fixed, thanks for pointing out